docs/usage/validators.md

```diff
@@ -1,375 +1,244 @@
+    assert False [type=assertion_error, input_value='John Doe!', input_type=str]
+    """
 ```
 
-## Reuse validators
-
-Occasionally, you will want to use the same validator on multiple fields/models (e.g. to
-normalize some input data). The "naive" approach would be to write a separate function,
-then call it from multiple decorators.  Obviously, this entails a lot of repetition and
-boiler plate code. To circumvent this, the `allow_reuse` parameter has been added to
-`pydantic.validator` in **v1.2** (`False` by default):
-
-```py
-from pydantic import BaseModel, field_validator
-
-
-def normalize(name: str) -> str:
-    return ' '.join((word.capitalize()) for word in name.split(' '))
-
-
-class Producer(BaseModel):
-    name: str
-
-    # validators
-    normalize_name = field_validator('name')(normalize)
-
-
-class Consumer(BaseModel):
-    name: str
-
-    # validators
-    normalize_name = field_validator('name')(normalize)
-
+A few things to note on validators:
 
-jane_doe = Producer(name='JaNe DOE')
-john_doe = Consumer(name='joHN dOe')
-assert jane_doe.name == 'Jane Doe'
-assert john_doe.name == 'John Doe'
-```
+* validators are "class methods", so the first argument value they receive is the `UserModel` class, not an instance of `UserModel`. We recommend you use the `@classmethod` decorator on them below the `@field_validator` decorator to get proper type checking.
+* the second argument is the field value to validate; it can be named as you please
+* the third argument, if present, is an instance of `pydantic.FieldValidationInfo`
+* validators should either return the parsed value or raise a `ValueError` or `AssertionError` (``assert`` statements may be used).
+* A single validator can be applied to multiple fields by passing it multiple field names.
+* A single validator can also be called on *all* fields by passing the special value `'*'`.
 
-As it is obvious, repetition has been reduced and the models become again almost
-declarative.
+!!! warning
+    If you make use of `assert` statements, keep in mind that running
+    Python with the [`-O` optimization flag](https://docs.python.org/3/using/cmdline.html#cmdoption-o)
+    disables `assert` statements, and **validators will stop working**.
 
-!!! tip
-    If you have a lot of fields that you want to validate, it usually makes sense to
-    define a help function with which you will avoid setting `allow_reuse=True` over and
-    over again.
+If you want to access values from another field you can use `FieldValidationInfo.data`, which is a dict of field name to field value.
+Validation is done in the order fields are defined, so you have to be careful when using `FieldValidationInfo.data` to not access a field that has not yet been validated/populated.
+In the example above you would not be able to access `info.data['id']` from within `name_must_contain_space`.
+For a more flexible approach to this we recommend you use `@model_validator`, which is discussed in the section below.
```

Comment:
I think it might be good to phrase this even more aggressively, something like:
```suggestion
If you want to access values from another field inside a `@field_validator`, this may be possible using `FieldValidationInfo.data`, which is a dict of field name to field value.
Validation is done in the order fields are defined, so you have to be careful when using `FieldValidationInfo.data` to not access a field that has not yet been validated/populated — 
in the code above, for example, you would not be able to access `info.data['id']` from within `name_must_contain_space`.
However, in most cases where you want to perform validation using multiple field values, it is better to use `@model_validator` which is discussed in the section below.
```
But I leave it up to you.